### ISSUE TITLE
`Pickles.compile` is accidentally async

### ISSUE THREAD

**Issue by L-as** [MinaProtocol/mina#16271]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/MinaProtocol/mina/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report. Please post those things in GitHub Discussions: https://github.com/MinaProtocol/mina/discussions

### Description

`Pickles.compile` returns before finishing compilation. This is evident by the "Snarky's internal state has been clobbered" error,
which happens I assume due to multiple circuits that use the impure interface being evaluated at the same time.

The workaround we employ is
```ocaml
let (tag, _, _, _) as r = Pickles.compile ... in
let _ = Async.Thread_safe.block_on_async_exn (fun () -> Pickles.Side_loaded.Verification_key.of_compiled tag) in
r
```

This works for us fine but I assume it's a symptom of some deeper bug somewhere.

Credit: @MartinOndejka found the issue originally IIRC.

### Steps to Reproduce

Described above.

### Expected Result

It should not need the caller to try to get the verification key for it to finish the compilation of the circuit.

### Actual Result

It doesn't finish compilation without waiting on it.

### Daemon version

Our fork begins at 7935f89cbe0461c570c4fbc9b456b96885e2cff3, but I assume the issue hasn't been fixed on the latest version of `compatible` yet.

### How frequently do you see this issue?

Always

### What is the impact of this issue on your ability to run a node?

Low

### Status

```shell
.
```
